weblab/entities/views.py

```diff
@@ -1014,3 +1014,73 @@ def get(self, request, *args, **kwargs):
         return JsonResponse({
             task: result
         })
+
+
+class EntityRunExperimentView(PermissionRequiredMixin, LoginRequiredMixin, EntityTypeMixin, DetailView):
+    """
+    Class for listing the possible experiment combinations
+    """
+    permission_required = 'experiments.create_experiment'
+    context_object_name = 'entity'
+    template_name = 'entities/entity_runexperiments.html'
+
+    def get_context_data(self, **kwargs):
+        entity = self.object
+
+        context = super().get_context_data(**kwargs)
+
+        # preposition to use in sentence: You may run this entity on/under the following entities
+        context['preposition'] = 'under'
+        if entity.entity_type == 'protocol':
+            context['preposition'] = 'on'
+
+        # ended up using a nested dict as nested lists caused django's unpacking in forloops to
+        # mess things up slightly
+        other_entities = Entity.objects.filter(entity_type=entity.other_type)\
+            .select_related('cachedentity').prefetch_related('cachedentity__versions', 'cachedentity__versions__tags')
+        context['object_list'] = []
+        context['other_object_list'] = []
+        for item in other_entities:
+            versions = item.cachedentity.versions
+            version_info = []
+            for version in versions.prefetch_related('tags'):
+                tag_list = list(version.tags.values_list('tag', flat=True))
+                commit = item.repo.get_commit(version.sha)
+                latest = item.repo.latest_commit
+                version_info.append({'commit': commit, 'tags': tag_list, 'latest': latest == commit})
+            if item.author == self.request.user:
+                context['object_list'].append({'id': item.id, 'name': item.name, 'versions': version_info})
+            else:
+                context['other_object_list'].append({'id': item.id, 'name': item.name, 'versions': version_info})
+        return context
+
+    def post(self, request, *args, **kwargs):
+        # this in not intuitive
+        # in get context self.object was the entity being worked with
+        # here we have to retrieve it
+        this_entity = self.get_object()
+
+        rerun = request.POST.get('rerun_expts')
+        exclude_existing = False
+        # look for 'None' string as this seems to be what test data passes for a NoneType
+        if not rerun or rerun == 'None':
+            exclude_existing = True
+        experiments_to_run = request.POST.getlist('runexperimentlist[]')
+        for version in experiments_to_run:
+            ident, sha = version.split(':')
+            exper_kwargs = {
+                this_entity.other_type + '_id': ident,
+                this_entity.other_type + '_version': sha,
+                this_entity.entity_type + '_id': this_entity.id,
+                this_entity.entity_type + '_version':  this_entity.repo.latest_commit.hexsha,
```

Comment:
Cache `this_entity.repo.latest_commit.hexsha` in a local variable outside the loop, otherwise you hit the filesystem every time around.

Reply:
fixed